pyteal/ast/opup.py

```diff
@@ -0,0 +1,164 @@
+    Note: the available budget must be high enough to execute the budget
+    increase code prior to invoking ensure_budget() or maximize_budget().
+
+    Example:
+        .. code-block:: python
+
+            opup = OpUp(OpUpMode.OnCall)
+            program_with_opup = Seq(
+                ...,
+                opup.ensure_budget(Int(1000)),
+                ...,
+            )
+    """
+
+    def __init__(self, mode: OpUpMode, target_app_id: Expr = None):
+        if mode == OpUpMode.Explicit:
+            if target_app_id is None:
+                raise TealInputError(
+                    "target_app_id must be specified in Explicit OpUp mode"
+                )
+            require_type(target_app_id, TealType.uint64)
+            self.target_app_id = target_app_id
+        elif mode == OpUpMode.OnCall:
+            self.target_app_id_slot = ScratchSlot()
+        else:
+            raise TealInputError("Invalid OpUp mode provided")
+
+        self.mode = mode
+
+        # A budget buffer is necessary to deal with an edge case of ensure_budget():
+        #   if the current budget is equal to or only slightly higher than the
+        #   required budget then it's possible for ensure_budget() to return with a
+        #   current budget less than the required budget. The buffer prevents this
+        #   from being the case.
+        self.buffer = Int(50)
+
+    def _create_app(self) -> Expr:
+        self.target_app_id = self.target_app_id_slot.load()
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.approval_program: ON_CALL_APP,
+                    TxnField.clear_state_program: ON_CALL_APP,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+            self.target_app_id_slot.store(Gitxn[0].created_application_id()),
+        )
+
+    def _delete_app(self, app_id: Expr) -> Expr:
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.on_completion: OnComplete.DeleteApplication,
+                    TxnField.application_id: app_id,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+    def _ensure_budget_expr(self, required_budget: Expr) -> Expr:
+        return While(required_budget < Global.opcode_budget()).Do(
+            Seq(
+                InnerTxnBuilder.Begin(),
+                InnerTxnBuilder.SetFields(
+                    {
+                        TxnField.type_enum: TxnType.ApplicationCall,
+                        TxnField.application_id: self.target_app_id,
+                    }
+                ),
+                InnerTxnBuilder.Submit(),
+            )
+        )
+
+    def ensure_budget(self, required_budget: Expr) -> Expr:
```

Comment:
I still don't see any docs here saying what the minimum budget should be before calling this function.

Reply:
Should be taken care of now.